File: 프로그래머스/Level2/2개 이하로 다른 비트.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<long long> solution(vector<long long> numbers) {
    vector<long long> answer;
    for(int i=0; i<numbers.size(); i++){
        if(numbers[i] % 2 == 0)
            answer.push_back(numbers[i] + 1);
        else{
            long long check = 1;
            while(true){
                if((numbers[i] & check) == 0)
                    break;
                check *= 2;
            }
            check /= 2;
            
            answer.push_back(numbers[i] + check);
        }
    }
    return answer;
}

File: 프로그래머스/Level2/2개 이하로 다른 비트.java
class Solution {
    public long[] solution(long[] numbers) {
        long[] answer = new long[numbers.length];
        for (int i = 0; i < numbers.length; ++i) {
            long number = numbers[i];
            
            long tmp = 1L;
            while ((number & tmp) != 0) {
                tmp <<= 1;
            }
            
            /* 가장 처음 나오는 0 -> 1 */
            number |= tmp;
            /* 0 오른쪽 1 -> 0 */
            number &= ~(tmp >> 1);
            
            answer[i] = number;
        }
        
        return answer;
    }
}
